### ISSUE TITLE
ROS1 tf2_ros conflicts and interference with ROS2 tf2_ros of header file transform_broadcaster.h 

### ISSUE THREAD

**Issue by gaoethan** [ros2/turtlebot2_demo#74]
## Bug report
When build the **turtlebot2_driver** package, its **kobuki_node** depends on **ROS1 kobuki_driver**, but after running `source /opt/ros/kinetic/setup.bash`, the building process is directed to use APIs in ROS1 tf2_ros transform_broadcaster.h, howerver, it should use those same APIs which are defined is ROS2 tf2_ros transform_broadcaster.h and have different arguments from ROS1. the following is the build error:

+++ Building 'turtlebot2_drivers'
==> '. /home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers/cmake__build.sh && /usr/bin/make cmake_check_build_system' in '/home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers'
==> '. /home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers/cmake__build.sh && /usr/bin/make -j4 -l4' in '/home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers'
[ 50%] Built target dumb_teleop
[ 75%] Building CXX object CMakeFiles/kobuki_node.dir/src/kobuki_node.cpp.o
/home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp: In function ‘int main(int, char**)’:
/home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp:89:40: error: no matching function for call to ‘tf2_ros::TransformBroadcaster::TransformBroadcaster(std::shared_ptr<rclcpp::Node>&)’
   tf2_ros::TransformBroadcaster br(node);
                                        ^
In file included from /home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp:42:0:
**/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:51:3: note: candidate: tf2_ros::TransformBroadcaster::TransformBroadcaster()
   TransformBroadcaster();
   ^
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:51:3: note:   candidate expects 0 arguments, 1 provided
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:48:7: note: candidate: tf2_ros::TransformBroadcaster::TransformBroadcaster(const tf2_ros::TransformBroadcaster&)
 class TransformBroadcaster{**
       ^
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:48:7: note:   no known conversion for argument 1 from ‘std::shared_ptr<rclcpp::Node>’ to ‘const tf2_ros::TransformBroadcaster&’
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:48:7: note: candidate: tf2_ros::TransformBroadcaster::TransformBroadcaster(tf2_ros::TransformBroadcaster&&)
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:48:7: note:   no known conversion for argument 1 from ‘std::shared_ptr<rclcpp::Node>’ to ‘tf2_ros::TransformBroadcaster&&’
/home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp:233:34: error: no matching function for call to ‘tf2_ros::TransformBroadcaster::sendTransform(geometry_msgs::msg::TransformStamped_<std::allocator<void> >&)’
     br.sendTransform(*odom_tf_msg);
                                  ^
In file included from /home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp:42:0:
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:63:8: note: candidate: void tf2_ros::TransformBroadcaster::sendTransform(const TransformStamped&)
   void sendTransform(const geometry_msgs::TransformStamped & transform);
        ^
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:63:8: note:   no known conversion for argument 1 from ‘geometry_msgs::msg::TransformStamped_<std::allocator<void> >’ to ‘const TransformStamped& {aka const geometry_msgs::TransformStamped_<std::allocator<void> >&}’
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:67:8: note: candidate: void tf2_ros::TransformBroadcaster::sendTransform(const std::vector<geometry_msgs::TransformStamped_<std::allocator<void> > >&)
   void sendTransform(const std::vector<geometry_msgs::TransformStamped> & transforms);
        ^
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:67:8: note:   no known conversion for argument 1 from ‘geometry_msgs::msg::TransformStamped_<std::allocator<void> >’ to ‘const std::vector<geometry_msgs::TransformStamped_<std::allocator<void> > >&’
/home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp:250:33: error: no matching function for call to ‘tf2_ros::TransformBroadcaster::sendTransform(geometry_msgs::msg::TransformStamped_<std::allocator<void> >&)’
     br.sendTransform(*imu_tf_msg);
                                 ^
In file included from /home/ros/Myspace/rosdev/ros2/release-latest/src/ros2/turtlebot2_demo/turtlebot2_drivers/src/kobuki_node.cpp:42:0:
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:63:8: note: candidate: void tf2_ros::TransformBroadcaster::sendTransform(const TransformStamped&)
   void sendTransform(const geometry_msgs::TransformStamped & transform);
        ^
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:63:8: note:   no known conversion for argument 1 from ‘geometry_msgs::msg::TransformStamped_<std::allocator<void> >’ to ‘const TransformStamped& {aka const geometry_msgs::TransformStamped_<std::allocator<void> >&}’
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:67:8: note: candidate: void tf2_ros::TransformBroadcaster::sendTransform(const std::vector<geometry_msgs::TransformStamped_<std::allocator<void> > >&)
   void sendTransform(const std::vector<geometry_msgs::TransformStamped> & transforms);
        ^
/opt/ros/kinetic/share/sophus/../../include/tf2_ros/transform_broadcaster.h:67:8: note:   no known conversion for argument 1 from ‘geometry_msgs::msg::TransformStamped_<std::allocator<void> >’ to ‘const std::vector<geometry_msgs::TransformStamped_<std::allocator<void> > >&’
CMakeFiles/kobuki_node.dir/build.make:62: recipe for target 'CMakeFiles/kobuki_node.dir/src/kobuki_node.cpp.o' failed
make[2]: *** [CMakeFiles/kobuki_node.dir/src/kobuki_node.cpp.o] Error 1
CMakeFiles/Makefile2:99: recipe for target 'CMakeFiles/kobuki_node.dir/all' failed
make[1]: *** [CMakeFiles/kobuki_node.dir/all] Error 2
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2

<== Command '. /home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers/cmake__build.sh && /usr/bin/make -j4 -l4' failed in '/home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers' with exit code '2'
<== Command '. /home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers/cmake__build.sh && /usr/bin/make -j4 -l4' failed in '/home/ros/Myspace/rosdev/ros2/release-latest/build/turtlebot2_drivers' with exit code '2'

**Required Info:**

- Operating System:
  -  Ubuntu 16.04
- Installation type:
  -  ROS2 source and ROS1 Binary
- Version or commit hash:
  -  master
- DDS implementation:
  -  N/A
- Client library (if applicable):
  -  N/A

#### Steps to reproduce issue

```
1. Clone the ROS2 release-latest (a.k.a ardent) and turtlebot2_demo ardent branch
2. Run `src/ament/ament_tools/scripts/ament.py build --build-tests --symlink-install --only-packages turtlebot2_drivers` after sourcing the ROS1 setup.bash which is installed in the way of binary package
```

#### Expected behavior
No build error and use ROS2 tf2_ros
#### Actual behavior
use ROS1 tf2_ros
#### Additional information

----

**Comment by clalancette** [comment]
Yeah, this is a known issue.  The build instructions [here](https://github.com/ros2/turtlebot2_demo/blob/master/README.md) very specifically split this up into two separate steps to avoid this issue (among others).  The long-term solution to this is to port `libkobuki.so` to ROS2, but it has a lot of dependencies, so that may take a while.

**Comment by gaoethan** [comment]
@clalancette thanks for the information, sorry, I'm a little confused for the workaround as you aforementioned:

> The build instructions here very specifically split this up into two separate steps to avoid this issue (among others). 

Do you mean these [two steps](https://github.com/ros2/turtlebot2_demo/blob/master/README.md#build-the-ros2-code) , but with my trial once it sources the ROS by "`source /opt/ros/kinetic/setup.bash`", then to build turtlebot2_drivers with the following cmd, this problem persists with the latest release (a. k. a ardent), thanks !

`src/ament/ament_tools/scripts/ament.py build --symlink-install  --only-packages turtlebot2_drivers`

**Comment by clalancette** [comment]
> Do you mean these two steps , but with my trial once it sources the ROS by "source /opt/ros/kinetic/setup.bash", then to build turtlebot2_drivers with the following cmd, this problem persists with the latest release (a. k. a ardent), thanks !

Yes, those are the correct steps.  You have to make sure that you do *not* source `/opt/ros/kinetic/setup.bash`, then build the core, and not the turtlebot* (and related) packages.  Once you have successfully done that, you source `/opt/ros/kinetic/setup.bash`, build just those turtlebot* (and related) packages.  If you are still having problems, make sure to completely log out of your terminal, then try following the directions again.  They seem to work for me.

**Comment by gaoethan** [comment]
okay, thanks, I removed the tf2_ros package from kinetic and then rebuilt the turtlebot2_drivers, it works, so   the key is that **no tf2_ros in kinetic should be involved**, or the conflict occurs.

**Comment by clalancette** [comment]
Great, glad it is working for you.  I'm actually going to close this one out, and open a different one suggesting that we port `libkobuki.so` to ROS2 to make this whole issue go away.
